### ISSUE TITLE
Weird time results when compiling with clang v6 and gcc 7+

### ISSUE THREAD

**Issue by ArashPartow** [beltoforion/muparser#68]
Weird time results when compiling with clang v6 and gcc 7+:

```
Expression 41 of 440: "((cos(cos(tan(pi)))+a)/a)"; Progress: #####
[01] ExprTk               (    8.000 ns,       1.491183914425581625,      13.683879358191861186)
[02] ExprTkFloat          (    9.000 ns,       1.491183876991271973,      13.683879375457763672)
[03] FParser 4.5          (   15.900 ns,       1.491183914425581625,      13.683879358191861186)
[04] muparser 2.2.4       (36478.800 ns,       1.491183914425581625,      13.683879358191861186)
[05] muparser 2.2.4 (omp) (38305.600 ns,       1.491183914425581625,      13.683879358191861186)
```

CPU details:
```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              20
On-line CPU(s) list: 0-19
Thread(s) per core:  2
Core(s) per socket:  10
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) W-2155 CPU @ 3.30GHz
Stepping:            4
CPU MHz:             1869.970
CPU max MHz:         4500.0000
CPU min MHz:         1200.0000
BogoMIPS:            6600.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            14080K
NUMA node0 CPU(s):   0-19

```

**Comment by beltoforion** [comment]
I don't really know what is going on there but my hunch is that the bytycode is not used. I can't really give you specifics without seeing the code.

**Comment by ArashPartow** [comment]
Here is the project:
[mathbenchmarks.zip](https://github.com/beltoforion/muparser/files/3992008/mathbenchmarks.zip)

Extract contents then to build execute the following:

`./build.sh`

To run the benchmark, execute the following:

`./ParserBenchmark 100000 benchmarks/bench_problematic.txt `

**Comment by ArashPartow** [comment]
Wondering if you've been able to confirm the previous results.

That being said, do let me know if you have any issues getting it build or run etc.

**Comment by beltoforion** [comment]
No, i did not investigate the issue. I do not have any indication that there is a general issue in muparser and my motivation to spend time fixing an issue that so far only the benchmark project has encountered is limited.

The benchmark project is massively biased towards exprtk. You essentially created lists with thousands or expressions tailormade for exprtk and mixed in a couple of hundret expressions with actual function calls from my old expression files. There seems to be an ominous almost complete lack of any multipliers for variables in the expression files. (Apart from the very few coming from my original files)

My goal when creating the original benchmark project was to provide a fair performance overview. The expressions used now consist mostly of thousands of variations of a use case where exprtk is good at.

For this reason i'm not too motivated to do any investigation into this specific issue as long as I do not get any feedback from other users of muparser that there is a problem. This however seems unlikely since the parsing engine of muparser wasn't change in years. So it may be something simple like the optimizer not being enabled.

**Comment by ArashPartow** [comment]
Ingo (@beltoforion ) I find your statements very confusing.

**Firstly** this was your benchmark suite, the code is entirely in your style *(camelcase, weird indentation)*, I have only been maintaining it for the past several years.

In fact since taking over I've only added three more parsers to your already available six parsers.


**Secondly** of the 9 benchmark files in the suite 4 of them are yours:

1. bench_expr.txt, bench_expr_all.txt,
2. bench_expr_random_with_functions.txt,
3. bench_expr_random_without_functions.txt


Two of the remaining *(bench_expr_weird.txt, bench_expr_precedence.txt)* I added in 2014, after we had a discussion over e-mail and where you were happy to have them as part of the benchmark and which you yourself added them to the repo back when this was still on Google code - I'm happy if needed to add the e-mail chains to this discussion to refresh your memory.


**Thirdly** the graph depicted on the page, are results from your own untouched benchmark *(bench_expr_random_without_functions.txt)*, again I can provide the e-mail chain(s) of the discussions where we agreed to this approach and to use the said particular benchmark file.


--------------


With regards to your comment about their being "biased" benchmarks - presumably you are talking about the benchmark files I've added:

1. bench_expr_complete.txt
2. bench_expr_all_permutations.txt
3. bench_expr_extensive.txt
4. bench_expr_precedence.txt
5. bench_expr_weird.txt

If you actually bothered to look at them, four of the five are all exhaustive combinations and permutations of all the base arithmetic operations between variables and literal values. There is clearly no bias of  one operation over another - If you think that is not the case please do let me know as I would like to fix this issue.

With regards to the last benchmark file **bench_expr_weird.txt** - this was added in consultation with you (again e-mails available). This particular benchmark was not intended for speed but rather correctness and If I remember correctly, you yourself fixed no less than two issues in muparser/muparser2 and muParserX based on this particular benchmark.

One thing you should note here is that the expression reported in this issue (mind you not the only one) actually comes from your very own **bench_expr_random_with_functions.txt** But that also not all the expression behave poorly either. If it was something across all the expressions that would be another issue entirely - but that is not the case.

Ingo you know you are always free to add any new benchmark files to the project as you see fit, I will always gladly and happily accept any of your contributions.


--------------

Ingo (@beltoforion ) I am very disappointed with your attitude and baseless accusations. You know very well we both set out with a goal to make an excellent, unbiased benchmark suite for C++ based mathematical expression parsers - to not only grade efficiency, but also accuracy and robustness of the parsers in the suite.

I have personally used the suite on countless occasions to verify changes to my own ExprTk library and I'm sure other people have made similar uses of the suite.

I have gone out of my way here to alert you to possible issues with your parser when being built with GCC and Clang on the Linux platform and was hoping to assist in correcting those issues as I am currently in the midst of rewriting the benchmark suite so that it can be portably compiled and run on both Windows and Linux OSes.

In short there are **no conspiracies** here, no one is out to get you, we are all here to try and help to maintain muParser as the excellent library it is.

So I say lets move on from the nonsensical accusations and statements of mistrust and lets fix whatever the issue is here with muParser and call it a day.

**Comment by beltoforion** [comment]
I have not been involved in any editorial decisions regarding the project for how long now 5 years? 8 years? I did not commit code since the move to GitHub. 

Had i been involved in any recent decisions I would have objected to removing the "Issues" tab from the math parser benchmark project. It is not a good look if the only public feedback channel is deactivated for a project claiming to be an "excellent, unbiased benchmark". Especially not if the only "officially" published chart is based on what I consider a biased expression set.

For the sake of transparency you should at least give every participant a chance to comment on the results. That is the very least i would expect from an "excellent, unbiased benchmark".

**Comment by ArashPartow** [comment]
>  published chart is based on what I consider a biased expression set

Ingo (@beltoforion) please read my comment:, it is **YOUR** benchmark **bench_expr_random_without_functions.txt** generated by using **FormelGenerator.cpp** committed by **YOU** when the project was on Google code *(e-mails can be pasted here for everyone to see if you like)*.

All those changes and additions I mentioned were done back as you said when the project was on Google code.


Ingo (@beltoforion) please move on from the nonsensical moronic accusations, and seriously have a look at the what I have reported.


Yes it is difficult as the author to get over the fact that you've spent soo much effort over such a long time working on muParser et al, and that much older less maintain libraries such as FParser and ATMSP (even ones where the authors have long since passed away) do much better in the benchmark results than your libraries - it's hard but it's also time to get over it.

--------------------

All I'm saying is give what I have presented above a try. Verify for yourself that there is something amiss here with muParser, verify that it is indeed being built with optimisations turned on and that any other compilation or runtime features you believe are needed are available.


It will only take you as long as decompressing that zip, building it and running it which should take less than a minute.


Don't be stubborn, either at best you'll find something wrong with what it is I'm doing and make me look the fool, at worst you'll have found an issue with muParser and will eventually fix it and make the library better for all its users.


So what do you have to loose?

**Comment by beltoforion** [comment]
I'm willing to explain my issues in detail on the issue tracker of the math parser benchmark project. Are you willing to discuss it there?

**Comment by ArashPartow** [comment]
Sure Ingo (@beltoforion ),  but lets first resolve the muParser. 

Have you had a chance to give the benchmark above a go? 

Have you been able to replicate the issue discussed?

--------------------------------------

Also feel free to make any PRs or fork the benchmark itslef.

**Comment by beltoforion** [comment]
The expressions that show this behavior all have pi in them right?

**Comment by ArashPartow** [comment]
It's possible that is the case  - but there's so many of them.  

At the same time there are many expressions muParser does well that  also has pi.

**Comment by beltoforion** [comment]
I cannot compile this with VS 2019 but its probebly this issue:
https://stackoverflow.com/questions/37504423/stdtan-extremely-slow-after-updating-glibc

**Comment by jschueller** [comment]
Ingo was right it is the libc tan that's slow for pi, if I cheat in muParserTemplateMagic.h:
```static T Tan(T v) { if(v==M_PI) return 0.0; return tan(v); }``` I've got 56ns which is more reasonable.
@ArashPartow How does exprtk evaluates tan ? I dont see where the functions are actually called.

**Comment by beltoforion** [comment]
I also found a reference in another old project that indicates the issue goes away if the accuracy of PI is reduced (https://github.com/beltoforion/muparser_experimental/blob/master/include/muParserMath.h bottom). The definition in muparser is using more than 15 decimal places which is more them MATH_PI is using. Simply taking the system value for PI should also eliminate this issue.

**Comment by jschueller** [comment]
I tried, using M_PI for PARSER_CONST_PI does not fix the benchmark:
```#define PARSER_CONST_PI M_PI```

**Comment by beltoforion** [comment]
Can you try specifically setting ti to this value:  3.141592653589? This appears to have fixed the issue the last time I tried. 

I'm planning to removed the macros and replace then with constants for the next service release but but currently i cannot commit because i have added a unit test that would break the build for GCC and CLANG.

**Comment by jschueller** [comment]
no, it does not fix it either

**Comment by ArashPartow** [comment]
Ingo @beltoforion   - I had a look at SO link. It seems the only person that seems to have a clue there is  Severin Pappadeux. His comment comes somewhat close to an explanation.

If for the tan call an avx/2 instruction is being emitted either by the compiler or is what is selected when a call is being made into libm either directly or via libc (runtime dispatch of the CPU), then that would explain it.

As due to thermal envelope issues most Intel (and AMD) processors will drastically scale down the applicable core for roughly 0.5-1sec for every avx applicable instruction that is executed on said core.

---------------------

@JSchueller

In this particular context, because Pi is a constant literal, ExprTk will perform a simple const folding optimisation leading to the cos(tan(pi)) part of the sub-expression being converted to a literal.

---------------------

Using the above denoted benchmark suite, I reran the benchmark with the following set of expressions:

    --- snip ---
    ((cos(cos(tan(pi)))+a)/a)
    tan(pi)
    tan(a)
    tan(3.14159265358979323846264338327950288419716939937511)
    tan(pi + a)
    tan(3.14159265358979323846264338327950288419716939937511 + a)
    cos(tan(pi))
    cos(tan(pi + a))
    cos(pi)
    cos(a)
    cos(3.14159265358979323846264338327950288419716939937511)
    cos(pi + a)
    cos(3.14159265358979323846264338327950288419716939937511 + a)
    sin(pi)/cos(pi)
    --- snip ---

Here are the muParser results:

    Expression 1 of 14: "((cos(cos(tan(pi)))+a)/a)"; Progress: ##
    [02] muparser 2.2.4 (41979.670 ns,       1.491183914425581625,   68419.396790920145576820)
    
    Expression 2 of 14: "tan(pi)"; Progress: ##
    [02] muparser 2.2.4 (42003.486 ns,      -0.000000000000000122,      -0.000000000006123234)
    
    Expression 3 of 14: "tan(a)"; Progress: ##
    [02] muparser 2.2.4 (   25.359 ns,       1.964759657248652314,   14773.415011996003158856)
    
    Expression 4 of 14: "tan(3.14159265358979323846264338327950288419716939937511)"; Progress: ##
    [02] muparser 2.2.4 (42324.630 ns,      -0.000000000000000122,      -0.000000000006123234)
    
    Expression 5 of 14: "tan(pi + a)"; Progress: ##
    [02] muparser 2.2.4 (   21.589 ns,       1.964759657248649649,   14773.415011996003158856)
    
    Expression 6 of 14: "tan(3.14159265358979323846264338327950288419716939937511 + a)"; Progress: ##
    [02] muparser 2.2.4 (   22.184 ns,       1.964759657248649649,   14773.415011996003158856)
    
    Expression 7 of 14: "cos(tan(pi))"; Progress: ##
    [02] muparser 2.2.4 (42591.673 ns,       1.000000000000000000,   50000.000000000000000000)
    
    Expression 8 of 14: "cos(tan(pi + a))"; Progress: ##
    [02] muparser 2.2.4 (   51.691 ns,      -0.383851139776528238,   -4703.727783833207467978)
    
    Expression 9 of 14: "cos(pi)"; Progress: ##
    [02] muparser 2.2.4 (   20.014 ns,      -1.000000000000000000,  -50000.000000000000000000)
    
    Expression 10 of 14: "cos(a)"; Progress: ##
    [02] muparser 2.2.4 (   23.505 ns,       0.453596121425577314,   -3372.624895745683716086)
    
    Expression 11 of 14: "cos(3.14159265358979323846264338327950288419716939937511)"; Progress: ##
    [02] muparser 2.2.4 (   19.961 ns,      -1.000000000000000000,  -50000.000000000000000000)
    
    Expression 12 of 14: "cos(pi + a)"; Progress: ##
    [02] muparser 2.2.4 (   21.893 ns,      -0.453596121425577814,    3372.624895745683261339)
    
    Expression 13 of 14: "cos(3.14159265358979323846264338327950288419716939937511 + a)"; Progress: ##
    [02] muparser 2.2.4 (   22.382 ns,      -0.453596121425577814,    3372.624895745683261339)
    
    Expression 14 of 14: "sin(pi)/cos(pi)"; Progress: ##
    [02] muparser 2.2.4 (  182.029 ns,      -0.000000000000000122,      -0.000000000006123234)
    

It seems whenever there is a non-const only input into tan, the behaviour is good, however when it is the opposite the problem arises.


--------------------------

Doing some debugging (note this is with O2), we get the following stack:

    [0] from 0x00007ffff76f5f9e in __tan_fma+78 at ../sysdeps/ieee754/dbl-64/s_tan.c:83
    [1] from 0x00000000008537f1 in mu::ParserBase::ParseCmdCodeBulk(int, int) const+2209 at /home/arash/workspace/temp/mathbenchmarks/src/muParser/muParserBase.cpp:1106
    [2] from 0x00000000008468b8 in BenchMuParser2::DoBenchmarkStd(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long)+3224 at /home/arash/workspace/temp/mathbenchmarks/src/BenchMuParser2.cpp:148
    [3] from 0x0000000000847f29 in Shootout(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::shared_ptr<Benchmark>, std::allocator<std::shared_ptr<Benchmark> > >&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, int, bool)+3209 at /home/arash/workspace/temp/mathbenchmarks/src/ParserBench.cpp:181
    [4] from 0x0000000000849137 in main(int, char const**)+1799 at /home/arash/workspace/temp/mathbenchmarks/src/ParserBench.cpp:453

The impl can be viewed here:
https://code.woboq.org/userspace/glibc/sysdeps/ieee754/dbl-64/s_tan.c.html#83

Can't be sure if this is what gets called when compiling with O3 (as debug symbols are not reliable) - would probably require disassembling.

**Comment by beltoforion** [comment]
Thanks for posting the results. I'll see what i can do but in the end i'll probably go with whatever libm or libc are doing. The issue may remain open if there is no easy fix. 

However muparser should also be doing constant folding so I don't see a reason for tan(pi) not being folded away unless pi was defined as a variable which i don't think it is. So this may be worthwhile to check out for me.
